Release 2.0 fixes

BethYates opened the pull request:
<!--
# sanger-tol/genomenote pull request

Many thanks for contributing to sanger-tol/genomenote!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/sanger-tol/genomenote/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/sanger-tol/genomenote/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

BethYates committed:
Rely on GBIF to provide correct Taxonomic authority information

BethYates committed:
Fixed rounding issues on genome/chr length

BethYates committed:
pre-process collection location to have same format across resources

BethYates committed:
GAL to title case

BethYates committed:
Lifestage in lower case

BethYates committed:
Added BUSCO data, fixed rounding issues and replaced 'tissue type' with 'organism part' for consistency across metadata sources

BethYates committed:
Don't strip parentheses from authority

BethYates committed:
Removed kingdom variable as it's not used directly

BethYates committed:
Ensure correct biosample accessions are set and check for "United Kingdom" in collection location

BethYates committed:
Update Higlass view config and add url to set of data to be passed to the genomenote template

BethYates committed:
minor fixes

BethYates committed:
black/prettier fixes

BethYates committed:
added alt_hap_accession to the set of parameters returned

BethYates committed:
New test species

BethYates committed:
black fix

BethYates committed:
minor fixes to parameter formating

BethYates committed:
add in percent assembled from goat

BethYates committed:
Don't output higlass link file - url now added to parameter files and GN document when applicable

BethYates committed:
Removed unnecessary metadata params and added checking of params provided

BethYates committed:
documentation updates

BethYates committed:
prettier fixes

BethYates committed:
linting fixes

BethYates committed:
No longer have params.species so need to set value in meta correctly

BethYates committed:
typo

BethYates committed:
changes required as no longer have params.species

BethYates committed:
updates ahead of release

muffato reviewed (commented):
I haven't completed the review but wanted to share what I have so far. I still need to go through all the `.nf`, the changelog, `docs/usage.md`, and `bin/check_parameters.py`.

tkchafin commented on bin/parse_json_copo_biosample.py:
```diff
@@ -99,6 +99,30 @@ def parse_json(file_in, file_out):
+                        location_list = param.split(" | ")
+                        location_list.reverse()
+
+                        # remove United Kingdom from location
+                        if "UNITED KINGDOM" in location_list:
+                            location_list.remove("UNITED KINGDOM")
+                        elif "United Kingdom" in location_list:
+                            location_list.remove("United Kingdom")
+
```
Maybe can convert to something like this, to catch cases that there is a case mistake like "UNited Kingdom" or "united kingdom"? Also added a `strip()` for any white space that might end up on the ends but I haven't looked to see if this was already done earlier in script

```
for loc in location_list:
    if loc.lower().strip() == "united kingdom":
        location_list.remove(loc)
```

BethYates replied:
I was removing it because only one of the data sources was including it so the location strings weren't matching across sources and the information wasn't making its way through to the populate template step. My thinking was it would be better to have some data than none and the note editor could add it back in if wanted.

BethYates replied:
Thinking about this is a better approach to compare arrays of all the location strings and if one array is longer than the others by one value but all other  array values match return the joined string of the longest array? Otherwise we're going to run into this issue for every possible country

muffato replied:
What about 1) making all the lists you get uniform by removing leading/trailing whitespace and lower case, 2) take the union of those sets into the genome note while reporting discrepancies in `genome_note_inconsistent.csv` ?
So even if there are discrepancies you'd still get something in the note.

tkchafin commented on bin/fetch_gbif_metadata.py:
```diff
@@ -91,7 +81,8 @@ def fetch_gbif_data(genus, species, output_file):
 def main(args=None):
     args = parse_args(args)
-    fetch_gbif_data(args.genus, args.species, args.output)
+    (genus, species) = args.species.split("_")
```
Could we do args.species.replace(" ", "_").split("_")? That would catch the case that species was input with a space as delimiter instead of underscore? I think those two would be the only cases we commonly use

BethYates replied:
species is not an input it's something we fetch in params_check.nf where we replace any spaces with underscores so I don't think this is necessary unless we want to have species as an optional pipeline input.

tkchafin committed:
Update CHANGELOG.md

tkchafin committed:
increase default memory for POPULATE_TEMPLATE

With 50MB I was having it fail and re-try

BethYates commented on bin/parse_json_ncbi_assembly.py:
```diff
@@ -135,6 +156,11 @@ def find_element(data, fields, attribs, param_list, index=0):
 
+        if "linked_assembly" in attribs.keys():
+            for assembly in data:
+                if "alternat" in assembly["assembly_type"] and "haplotype" in assembly["assembly_type"]:
+                    return assembly["linked_assembly"]
```
I'll remove it.

BethYates commented on bin/populate_genome_note_template.py:
```diff
@@ -84,6 +94,12 @@ def build_param_list(param_file):
         mydict["AUTHORS"] = ", ".join(authors)
 
+        if mydict["ASSEMBLY_ACCESSION"] and mydict["GENUS_SPECIES"]:
+            btk_busco_url = "https://blobtoolkit.genomehubs.org/view/SPECIES/dataset/GCA/busco"
```
change made

muffato commented:
I got this error in `UPLOAD_HIGLASS_DATA` when running the full test:
```
Command error:
  W1009 12:03:04.412509      41 loader.go:222] Config not found: /nfs/users/nfs_m/mm49/.kube/config.tol-it-dev-k8s
  W1009 12:03:04.450610      96 loader.go:222] Config not found: /nfs/users/nfs_m/mm49/.kube/config.tol-it-dev-k8s
  error: no current context is set
```
I can see the script does
```
export KUBECONFIG=~/.kube/config.tol-it-dev-k8s
```
and the file exists. I guess it doesn't work because the file is not given as an input file to the module so Nextflow cannot stage it properly, and the way I run the pipeline `/nfs` is not mounted automatically either.

Running the full test with `--upload_higlass_data false` works.

Given this is not needed for BGA, let's push that to the 2.0.1 release ?

muffato commented on nextflow_schema.json:
Update the `required` list to include `assembly` and `biosample_wgs`

muffato commented on subworkflows/local/input_check.nf:
```diff
@@ -2,35 +2,81 @@
+    // set temp key to allow combining channels
+    param
+        .map { meta ->
+            [meta.id[0], meta]
```
I'm confused by `[0]`. Isn't `meta.id` the accession number `GCA_0123224.1` ?

BethYates replied:
When I used just meta.id  rather than meta.id[0] when I ran  a .view() on the output channel the channel looked like [ [GCA_0123224.1], [ meta....] ..]  I had to use meta.id[0] to get the accession number as a value not a list. I'm not sure why, unless theres something wrong with the way I'm setting up the param channel

muffato replied:
I think it got to do with the `param` channel getting `[meta]` instead of `meta`. It means `param` is actually like this
```
[[id:GCA_963859965.1, species:Ceramica_pisi, taxon_id:988087, bioproject:PRJEB65408, biosample_wgs:SAMEA112198456, biosample_hic:SAMEA112198479, biosample_rna:SAMEA112232914]]
```
(note the double brackets). Then `ch_tmp_param` is:
```
[GCA_963859965.1, [[id:GCA_963859965.1, species:Ceramica_pisi, taxon_id:988087, bioproject:PRJEB65408, biosample_wgs:SAMEA112198456, biosample_hic:SAMEA112198479, biosample_rna:SAMEA112232914]]]
```

I don't understand why `meta.id[0]` works, naturally I'd have expected `meta[0].id` in this case. I quickly tried to use just `meta`, change the `[0]` etc, but then the pipeline crashes for me. Given the little time we have before BGA and the fact the tests work, let's just not touch this for now ?

muffato commented on CHANGELOG.md:
```diff
@@ -3,6 +3,36 @@
+|               | --higlass_kubeconfig       |
+|               | --higlass_upload_directory |
+|               | --higlass_data_project_dir |
+
```
```suggestion
| --taxon_id    |                            |
| --species     |                            |
| --bioproject  |                            |

```

BethYates replied:
These params were never actually on the dev or release branch so I assumed that we don't need to mention them?

muffato replied:
Gotcha. You're right.

muffato commented on bin/check_parameters.py:
All your `if response.status_code == 200:` are missing their `else` counterpart, meaning the output CSV may have different columns. Either raise an error or output `null` in the CSV

muffato commented on bin/check_parameters.py:
```diff
@@ -0,0 +1,128 @@
+            header.append("hic_biosample")
+            params.append("null")
+
+        if rna_biosample and rna_biosample != "null":
```
```suggestion
        # Validate rna_biosample
        if rna_biosample and rna_biosample != "null":
```

muffato commented on bin/check_parameters.py:
```diff
@@ -0,0 +1,128 @@
+        params = [assembly, species, taxon_id]
+        header = ["assembly", "species", "taxon_id"]
+
+        if study:
```
This `if` is also missing an `else`

muffato commented on bin/check_parameters.py:
```diff
@@ -0,0 +1,128 @@
+                params.append(wgs_biosample)
+                header.append("wgs_biosample")
+
+        if hic_biosample and hic_biosample != "null":
```
```suggestion
        # Validate hic_biosample
        if hic_biosample and hic_biosample != "null":
```

muffato reviewed (commented):
Also found a minor issue in the resulting note that can be addressed in 2.0.1. The collector name should be capitalised in
```
The specimen was collected by DOUGLAS BOYES (University of Oxford) and identified by DOUGLAS BOYES (University of Oxford) and preserved by dry ice/coolrack. 
``` 

Oh, I remembered that the issue with `--upload_higlass_data` is going to break the full test run after release that is used to provide sample output data for the pipelines website. Perhaps turn the option off in 2.0.0 and reenable it in 2.0.1 ?

And that's all for me ! I can confirm I can run the pipeline on both test profiles and I get beautiful genome notes.

tkchafin committed:
remove duplicate container directive

first directive is overwritten by second one, have tested that 2nd is correct! :)

tkchafin committed:
Change to same container as other requests module

One less container to fetch; have tested

BethYates committed:
Don't upload to Higlass in full test

BethYates committed:
used elif rather than lots of ifs in parsing scripts, handle collection_location parsing/matching better and include all linked assemblies

BethYates committed:
Merge branch 'release_2.0_fixes' of github.com:sanger-tol/genomenote into release_2.0_fixes

BethYates committed:
black fix

BethYates commented on subworkflows/local/genome_metadata.nf:
```diff
@@ -12,55 +12,52 @@ include { FETCH_GBIF_METADATA       }       from '../../modules/local/fetch_gbif
-            type: row.type,
-            ext: row.ext
-        ]
+        def new_meta = metadata.clone()
+        new_meta.source = row.source
+        new_meta.type = row.type
+        new_meta.ext = row.ext
```
cloning the metadata was the first way I got this to work - trying the other approaches you suggested was failing as I wasn't assigning the modified metadata to anything so the new data wasn't being passed through properly - I've fixed this now.

BethYates committed:
fixed extending meta

BethYates committed:
add required parameters

BethYates committed:
simplify meta setting

BethYates committed:
added comments and appropriate error raising

BethYates committed:
Moved sanger specific running instructions to the end of the file

BethYates committed:
prettier fix

BethYates commented on docs/usage.md:
````diff
@@ -37,16 +34,16 @@ If you wish to run the optional step that writes genome metatdata key value-pair
 ```bash
-  nextflow secrets set TOL_API_KEY '[API key]'
+   --note_template '[URL for Genome Notes Portal API]'
 ```
````
Moved to the end of the file

BethYates committed:
Fix KeyError Exceptions and ensure getting all biosample collectors/identifiers for the author string

BethYates committed:
black fixes

BethYates commented:
> Also found a minor issue in the resulting note that can be addressed in 2.0.1. The collector name should be capitalised in
> 
> ```
> The specimen was collected by DOUGLAS BOYES (University of Oxford) and identified by DOUGLAS BOYES (University
> ```

I think this issue has been fixed by the latest changes I made to the populate_genome_notes_template.py script - the changes were necessary to fix a KeyError bug I found when running the pipeline on my wider test species

muffato reviewed (approved):
Good for me !

By the way, minor thing, on the full test, It still says "University **Of** Oxford" even though you've got code to turn `Of` to `of`. Do you know why it's still happening ?
>  Douglas Boyes, $INTRO_WRITER, University Of Oxford Genome Acquisition Lab, University Of Oxford Genome Acquisition Lab, Darwin Tree of Life Barcoding collective, Wellcome Sanger Institute Tree of Life Management, Samples and Laboratory team, Wellcome Sanger Institute Scientific Operations: Sequencing Operations, Wellcome Sanger Institute Tree of Life Core Informatics team, Tree of Life Core Informatics collective, Darwin Tree of Life Consortium

BethYates commented:
> Good for me !
> 
> By the way, minor thing, on the full test, It still says "University **Of** Oxford" even though you've got code to turn `Of` to `of`. Do you know why it's still happening ?
> 
> > Douglas Boyes, $INTRO_WRITER, University Of Oxford Genome Acquisition Lab, University Of Oxford Genome Acquisition Lab, Darwin Tree of Life Barcoding collective, Wellcome Sanger Institute Tree of Life Management, Samples and Laboratory team, Wellcome Sanger Institute Scientific Operations: Sequencing Operations, Wellcome Sanger Institute Tree of Life Core Informatics team, Tree of Life Core Informatics collective, Darwin Tree of Life Consortium

I'm not sure, I noticed this yesterday too but didn't have time to investigate. There's probably lots of other template changes we could make to how we handle authors and their affiliations as well as collector/identifier information in the methods section so we could address this when we look at that. This is something I've been talking to Karen about and I want to get the logic right with her before doing anything more.
